papers/hatra21/talk.tex

```diff
@@ -0,0 +1,203 @@
+  \item \textbf{Type-driven tools} are useful, e.g.~autocomplete or API help.
+  \item \textbf{Type errors} may be useful (e.g.~catching typos) but some are not.
+\end{itemize}
+Type systems should help or get out of the way.
+
+\end{frame}
+
+\begin{frame}
+
+\frametitle{Professionals have long-term goals}
+
+E.g. ``decrease user churn'' or ``improve frame rate''.
+\begin{itemize}
+\item \textbf{Code planning}: programs are incomplete.
+\item \textbf{Code refactoring}: programs change.
+\item \textbf{Defect detection}: programs have bugs.
+\end{itemize}
+Type-driven development is a useful technique!
+
+\end{frame}
+
+\part{Luau Type System}
+
+\begin{frame}
+
+\frametitle{Infallible types}
+
+Goal: \emph{support type-driven tools (e.g. autocomplete) for all programs.}
+\begin{itemize}
+\item \textbf{Traditional typing judgment} says nothing about ill-typed terms.
+\item \textbf{Infallible judgment}: every term gets a type.
+\item \textbf{Flag type errors}: elaboration introduces \emph{flagged} subterms.
+\end{itemize}
+
+\emph{Related work}:
+\begin{itemize}
+\item Type error reporting, program repair.
+\item Typed holes (e.g. in Hazel).
+\end{itemize}
+
+\end{frame}
+
+\begin{frame}
+
+\frametitle{Strict types}
+
+Goal: \emph{no false negatives}.
+
+\begin{itemize}
+\item \textbf{Strict mode} enabled by developers who want defect detection.
+\item \textbf{Business as usual} soundness via progress + preservation.
+\item \textbf{Gradual types} for programs with flagged type errors.
+\end{itemize}
+
+\emph{Related work}:
+\begin{itemize}
+\item Lots and lots for type safety.
+\item Gradual typing, blame analysis, migratory types\dots
+\end{itemize}
+
+\end{frame}
+
+\begin{frame}
+
+\frametitle{Nonstrict types}
+
+Goal: \emph{no false positives}.
+
+\begin{itemize}
+\item \textbf{Nonstrict mode} enabled by developers who want type-drive tools.
+\item \textbf{Victory condition} does not have an obvious definition!
+\item \textbf{A shot at it}: a program is \emph{incorrectly flagged} if it contains
+  a flagged value (i.e.~a flagged program has successfully terminated).
+\item \textbf{Progress + correct flagging} is what we want???
+\end{itemize}
+
+\emph{Related work}:
+\begin{itemize}
+\item Success types (e.g. Erlang Dialyzer).
+\item Incorrectness Logic.
+\end{itemize}
+
+\end{frame}
+
+\begin{frame}
+
+\frametitle{Mixing types}
+
+Goal: \emph{support mixed strict/nonstrict development}.
+
+\begin{itemize}
+\item \textbf{Per-module} strict/nonstrict mode.
+\item \textbf{Combined} progress + preservation with progress + correct flagging?
+\end{itemize}
+
+\emph{Related work}:
+\begin{itemize}
+\item Some on mixed languages, but with shared safety properties.
+\end{itemize}
+  
+\end{frame}
+
+\begin{frame}
+
+\frametitle{Type inference}
+
+Goal: \emph{provide benefits of type-directed tools to everyone}.
+
+\begin{itemize}
+\item \textbf{Infer types} for all variables, don't just give them type any.
```

Comment:
We should phrase this in a way that makes sense to people who have never used TypeScript or Luau.  `any` won't make sense to those people.

How about this?
```suggestion
\item \textbf{Infer types} for all variables.  Resist the urge to give up and ascribe a top type when an error is encountered.
```